Function can not return chrono::DateTime<Local> since 0.25.0

### Bug Description

I tried upgrading my pyo3 version to 0.25.0 but my code won't compile anymore on functions that return a Option<DateTime<Local>> from the chrono crate.


### Steps to Reproduce

1. clone example repo at: [https://github.com/BenediktMaag/pyo3-return-error](https://github.com/BenediktMaag/pyo3-return-error/tree/main)
2. run cargo build (will fail)
3. cargo add pyo3@0.24.2 --features chrono
4. run cargo build (will succeed)

### Backtrace

```shell
Compiling pyo3-return-error v0.1.0 (C:\temp\pyo3-return-error)
error[E0277]: `Local` cannot be converted to a Python object
   --> src\lib.rs:5:23
    |
5   | fn get_timestamp() -> Option<DateTime<Local>> {
    |                       ^^^^^^ the trait `IntoPyObject<'_>` is not implemented for `Local`
    |
    = note: `IntoPyObject` is automatically implemented by the `#[pyclass]` macro
    = note: if you do not wish to have a corresponding Python type, implement it manually
    = note: if you do not own `Local` you can perform a manual conversion to one of the types in `pyo3::types::*`
    = help: the following other types implement trait `IntoPyObject<'py>`:
              &&'a T
              &&OsStr
              &&Path
              &&str
              &'a (T0, T1)
              &'a (T0, T1, T2)
              &'a (T0, T1, T2, T3)
              &'a (T0, T1, T2, T3, T4)
            and 152 others
    = note: required for `DateTime<Local>` to implement `IntoPyObject<'_>`
    = note: 1 redundant requirement hidden
    = note: required for `Option<DateTime<Local>>` to implement `IntoPyObject<'_>`
note: required by a bound in `UnknownReturnType::<T>::wrap`
   --> C:\Users\MaagB\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\pyo3-0.25.0\src\impl_\wrap.rs:123:12
    |
121 |     pub fn wrap<'py>(&self, _: T) -> T
    |            ---- required by a bound in this associated function
122 |     where
123 |         T: IntoPyObject<'py>,
    |            ^^^^^^^^^^^^^^^^^ required by this bound in `UnknownReturnType::<T>::wrap`

error[E0599]: no method named `map_err` found for enum `Option` in the current scope
    --> src\lib.rs:5:23
     |
5    | fn get_timestamp() -> Option<DateTime<Local>> {
     |                       ^^^^^^
     |
help: there is a method `map_or` with a similar name, but with different arguments
    --> C:\Users\MaagB\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\core\src\option.rs:1173:5
     |
1173 | /     pub fn map_or<U, F>(self, default: U, f: F) -> U
1174 | |     where
1175 | |         F: FnOnce(T) -> U,
     | |__________________________^

Some errors have detailed explanations: E0277, E0599.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `pyo3-return-error` (lib) due to 2 previous errors
```

### Your operating system and version

Windows 11

### Your Python version (`python --version`)

Python 3.12

### Your Rust version (`rustc --version`)

1.86.0

### Your PyO3 version

0.25.0

### How did you install python? Did you use a virtualenv?

uv / uv run maturin develop

### Additional Info

_No response_

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -51,6 +51,7 @@ smallvec = { version = "1.0", optional = true }
 uuid = { version = "1.11.0", optional = true }
 lock_api = { version = "0.4", optional = true }
 parking_lot = { version = "0.12", optional = true }
+iana-time-zone = { version = "0.1", optional = true, features = ["fallback"]}
 
 [target.'cfg(not(target_has_atomic = "64"))'.dependencies]
 portable-atomic = "1.0"
@@ -121,6 +122,8 @@ py-clone = []
 parking_lot = ["dep:parking_lot", "lock_api"]
 arc_lock = ["lock_api", "lock_api/arc_lock", "parking_lot?/arc_lock"]
 
+chrono-local = ["chrono/clock", "dep:iana-time-zone"]
+
 # Optimizes PyObject to Vec conversion and so on.
 nightly = []
 
@@ -133,6 +136,7 @@ full = [
     "arc_lock",
     "bigdecimal",
     "chrono",
+    "chrono-local",
     "chrono-tz",
     "either",
     "experimental-async",

--- a/guide/src/features.md
+++ b/guide/src/features.md
@@ -138,6 +138,10 @@ Adds a dependency on [chrono](https://docs.rs/chrono). Enables a conversion from
 - [NaiveTime](https://docs.rs/chrono/latest/chrono/naive/struct.NaiveTime.html) -> [`PyTime`]({{#PYO3_DOCS_URL}}/pyo3/types/struct.PyTime.html)
 - [DateTime](https://docs.rs/chrono/latest/chrono/struct.DateTime.html) -> [`PyDateTime`]({{#PYO3_DOCS_URL}}/pyo3/types/struct.PyDateTime.html)
 
+### `chrono-local`
+
+Enables conversion from and to [Local](https://docs.rs/chrono/latest/chrono/struct.Local.html) timezones.
+
 ### `chrono-tz`
 
 Adds a dependency on [chrono-tz](https://docs.rs/chrono-tz).

--- a/newsfragments/5174.added.md
+++ b/newsfragments/5174.added.md
@@ -1,0 +1,1 @@
+Add conversions for chrono's `Local` timezone & `DateTime<Local>` instances.

--- a/src/conversions/chrono.rs
+++ b/src/conversions/chrono.rs
@@ -43,15 +43,23 @@
 
 use crate::conversion::IntoPyObject;
 use crate::exceptions::{PyTypeError, PyUserWarning, PyValueError};
-#[cfg(Py_LIMITED_API)]
 use crate::intern;
 use crate::types::any::PyAnyMethods;
 use crate::types::PyNone;
 use crate::types::{PyDate, PyDateTime, PyDelta, PyTime, PyTzInfo, PyTzInfoAccess};
 #[cfg(not(Py_LIMITED_API))]
 use crate::types::{PyDateAccess, PyDeltaAccess, PyTimeAccess};
+#[cfg(feature = "chrono-local")]
+use crate::{
+    exceptions::PyRuntimeError,
+    sync::GILOnceCell,
+    types::{PyString, PyStringMethods},
+    Py,
+};
 use crate::{ffi, Borrowed, Bound, FromPyObject, IntoPyObjectExt, PyAny, PyErr, PyResult, Python};
 use chrono::offset::{FixedOffset, Utc};
+#[cfg(feature = "chrono-local")]
+use chrono::Local;
 use chrono::{
     DateTime, Datelike, Duration, LocalResult, NaiveDate, NaiveDateTime, NaiveTime, Offset,
     TimeZone, Timelike,
@@ -387,7 +395,8 @@ impl FromPyObject<'_> for FixedOffset {
         // Any other timezone would require a datetime as the parameter, and return
         // None if the datetime is not provided.
         // Trying to convert None to a PyDelta in the next line will then fail.
-        let py_timedelta = ob.call_method1("utcoffset", (PyNone::get(ob.py()),))?;
+        let py_timedelta =
+            ob.call_method1(intern!(ob.py(), "utcoffset"), (PyNone::get(ob.py()),))?;
         if py_timedelta.is_none() {
             return Err(PyTypeError::new_err(format!(
                 "{ob:?} is not a fixed offset timezone"
@@ -429,6 +438,54 @@ impl FromPyObject<'_> for Utc {
             Ok(Utc)
         } else {
             Err(PyValueError::new_err("expected datetime.timezone.utc"))
+        }
+    }
+}
+
+#[cfg(feature = "chrono-local")]
+impl<'py> IntoPyObject<'py> for Local {
+    type Target = PyTzInfo;
+    type Output = Borrowed<'static, 'py, Self::Target>;
+    type Error = PyErr;
+
+    fn into_pyobject(self, py: Python<'py>) -> Result<Self::Output, Self::Error> {
+        static LOCAL_TZ: GILOnceCell<Py<PyTzInfo>> = GILOnceCell::new();
+        let tz = LOCAL_TZ
+            .get_or_try_init(py, || {
+                let iana_name = iana_time_zone::get_timezone().map_err(|e| {
+                    PyRuntimeError::new_err(format!("Could not get local timezone: {e}"))
+                })?;
+                PyTzInfo::timezone(py, iana_name).map(Bound::unbind)
+            })?
+            .bind_borrowed(py);
+        Ok(tz)
+    }
+}
+
+#[cfg(feature = "chrono-local")]
+impl<'py> IntoPyObject<'py> for &Local {
+    type Target = PyTzInfo;
+    type Output = Borrowed<'static, 'py, Self::Target>;
+    type Error = PyErr;
+
+    #[inline]
+    fn into_pyobject(self, py: Python<'py>) -> Result<Self::Output, Self::Error> {
+        (*self).into_pyobject(py)
+    }
+}
+
+#[cfg(feature = "chrono-local")]
+impl FromPyObject<'_> for Local {
+    fn extract_bound(ob: &Bound<'_, PyAny>) -> PyResult<Local> {
+        let local_tz = Local.into_pyobject(ob.py())?;
+        if ob.eq(local_tz)? {
+            Ok(Local)
+        } else {
+            let name = local_tz.getattr("key")?.downcast_into::<PyString>()?;
+            Err(PyValueError::new_err(format!(
+                "expected local timezone {}",
+                name.to_cow()?
+            )))
         }
     }
 }
@@ -1278,6 +1335,43 @@ mod tests {
                         let expected_roundtrip_time = micro.checked_sub(1_000_000).map(|micro| NaiveTime::from_hms_micro_opt(hour, min, sec, micro).unwrap()).unwrap_or(time);
                         let expected_roundtrip_dt: DateTime<FixedOffset> = NaiveDateTime::new(date, expected_roundtrip_time).and_local_timezone(offset).unwrap();
                         assert_eq!(expected_roundtrip_dt, roundtripped);
+                    }
+                })
+            }
+
+            #[test]
+            #[cfg(all(feature = "chrono-local", not(target_os = "windows")))]
+            fn test_local_datetime_roundtrip(
+                year in 1i32..=9999i32,
+                month in 1u32..=12u32,
+                day in 1u32..=31u32,
+                hour in 0u32..=23u32,
+                min in 0u32..=59u32,
+                sec in 0u32..=59u32,
+                micro in 0u32..=1_999_999u32,
+            ) {
+                Python::with_gil(|py| {
+                    let date_opt = NaiveDate::from_ymd_opt(year, month, day);
+                    let time_opt = NaiveTime::from_hms_micro_opt(hour, min, sec, micro);
+                    if let (Some(date), Some(time)) = (date_opt, time_opt) {
+                        let dts = match NaiveDateTime::new(date, time).and_local_timezone(Local) {
+                            LocalResult::None => return,
+                            LocalResult::Single(dt) => [Some((dt, false)), None],
+                            LocalResult::Ambiguous(dt1, dt2) => [Some((dt1, false)), Some((dt2, true))],
+                        };
+                        for (dt, fold) in dts.iter().filter_map(|input| *input) {
+                            // Wrap in CatchWarnings to avoid into_py firing warning for truncated leap second
+                            let py_dt = CatchWarnings::enter(py, |_| dt.into_pyobject(py)).unwrap();
+                            let roundtripped: DateTime<Local> = py_dt.extract().expect("Round trip");
+                            // Leap seconds are not roundtripped
+                            let expected_roundtrip_time = micro.checked_sub(1_000_000).map(|micro| NaiveTime::from_hms_micro_opt(hour, min, sec, micro).unwrap()).unwrap_or(time);
+                            let expected_roundtrip_dt: DateTime<Local> = if fold {
+                                NaiveDateTime::new(date, expected_roundtrip_time).and_local_timezone(Local).latest()
+                            } else {
+                                NaiveDateTime::new(date, expected_roundtrip_time).and_local_timezone(Local).earliest()
+                            }.unwrap();
+                            assert_eq!(expected_roundtrip_dt, roundtripped);
+                        }
                     }
                 })
             }

